agent/integration/job_runner_integration_test.go

```diff
@@ -4,14 +4,117 @@ import (
 	"context"
 	"fmt"
 	"os"
+	"path/filepath"
+	"runtime"
 	"strconv"
 	"testing"
 
 	"github.com/buildkite/agent/v3/agent"
 	"github.com/buildkite/agent/v3/api"
 	"github.com/buildkite/bintest/v3"
+	"gotest.tools/v3/assert"
 )
 
+func TestPreBootstrapHookScripts(t *testing.T) {
+	t.Parallel()
+
+	type testCase struct {
+		name     string
+		contents string
+		ext      string
+		allowed  bool
+	}
+
+	testCases := []testCase{
+		{
+			name:     "sh_success",
+			contents: "#!/bin/sh\necho hello from a pre-bootstrap hook!\n",
+			ext:      "",
+			allowed:  true,
+		},
+		{
+			name:     "sh_failure",
+			contents: "#!/bin/sh\nexit 1\n",
+			ext:      "",
+			allowed:  false,
+		},
+	}
+
+	if runtime.GOOS == "windows" {
+		testCases = append(
+			testCases,
+			testCase{
+				name:     "bat_success",
+				contents: "exit 0",
+				ext:      ".bat",
+				allowed:  true,
+			},
+			testCase{
+				name:     "bat_failure",
+				contents: "exit 1",
+				ext:      ".bat",
+				allowed:  false,
+			},
+			testCase{
+				name:     "powershell_failure",
+				contents: "Exit 0",
+				ext:      ".ps1",
+				allowed:  true,
+			},
+			testCase{
+				name:     "powershell_failure",
+				contents: "Exit 1",
+				ext:      ".ps1",
+				allowed:  false,
+			},
+		)
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
```

Comment:
I think `RunScript` also runs scripts with shebang lines directly, so might this occasionally flake with the ETXTBSY bug?

Reply:
Good point, I'll write the hook in a subprocess.

Reply:
It's written by a subprocess now. PTAL

Reply:
Hmmm... hang on a sec...

The scripts are all very short, and don't change. Wouldn't it be less work to make them test fixtures?

Reply:
Considering I've already done the work, is this moot?

Reply:
I suppose so. If anything I just regret not having the realisation 5 days ago that it could have been implemented with a handful of files checked into the repo. 

Though I suppose this also has to be maintained... no, not going to argue further. It's fine.